File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    pair<int, bool> f(TreeNode* root) {
        if(!root) {
            return {0, true};
        }
        
        auto l = f(root->left);
        auto r = f(root->right);
        
        if(!l.second || !r.second) {
            return {-1, false};
        }
        
        
        return {max(l.first, r.first)+1, (abs(l.first-r.first) <= 1)};
    }
    bool isBalanced(TreeNode* root) {
        return f(root).second;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import ("math")

func f(root *TreeNode) (float64, bool) {
    if root == nil {
        return 0, true
    }
    
    lh, ok := f(root.Left)
    if !ok {
        return 0, false
    }
    
    rh, ok := f(root.Right)
    if !ok || (math.Abs(lh-rh) > 1) {
        return 0, false;
    }
    
    return math.Max(lh, rh)+1, true
}

func isBalanced(root *TreeNode) bool {
    _, res := f(root)
    return res
}